### ISSUE TITLE
Take over try.haxe.org by HF

### ISSUE THREAD

**Issue by nadako** [HaxeFoundation/Project-Management#58]
Hey people.

Since http://try.haxe.org/ is currently down (ping @clemos) and there's a new unmerged development of tryhaxe supporting macros by @mrcdk (http://try-haxe.mrcdk.com/), I thought how about we take the project under the HF's wing to provide official support and development for it.

Right now I feel like it looks kind of weak that the thing that seems to be pretty official (e.g. it's under the haxe.org domain) is not really developed and noone from HF is responsible for it being broken. This is a good resource that we should take care of.

I'm not sure if we have the server management resources for that, since @andyli is pretty busy with other stuff, but I think I could help with the further development. For example I was thinking that we could improve the editing experience by integrating this new [Monaco](https://microsoft.github.io/monaco-editor/) editor that powers VSCode and the [Haxe Language Server](https://github.com/vshaxe/haxe-languageserver) so we have fancier completion, hover helps, goto definitions and other stuff.

**Comment by clemos** [comment]
Hi all,
As I already said to @mrcdk, I would be happy to let someone else take care of try.haxe.org, and would definitely give a hand for an eventual transfer.
I don't have much time to maintain the server any more (which is up again, sorry :), even less for new features, and I also agree it would make sense for HF to officially support it (@markknol has already taken over the design for quite a while ;) ).

**Comment by andyli** [comment]
@mrcdk had contacted me about that he could handle the server and maintenance of it last week.

@ncannasse: I've mentioned that on slack just before your holiday but it seems it didn't reach you. @mrcdk want to know if HF can cover the server cost in case it is too high (>80$/month) for him.

@mrcdk: Have you tried the free tier of IBM's Bluemix yet?

**Comment by clemos** [comment]
FYI, I pay ~30€/month for the current try.haxe.org server. This is for the whole dedicated server, which also hosts some other projects of mine. 
So it's pretty difficult to evaluate, but I would say that with half that cost, you would already have something that would easily support the load.
Macro / sys support (Docker) would probably require a bit more resource, but currently the server is very far from being overloaded.

**Comment by ncannasse** [comment]
I don't expect the server costs to be too high. In case they do, then HF can provide a server for hosting, which we will pay and eventually use for other services as well.

**Comment by ibilon** [comment]
Considering that the issue is that tryhaxe isn't official enough/not officially support by HF,
personally I think that not having it on a HF server is weird.

**Comment by nadako** [comment]
yeah, it should be moved to our servers/clouds and maintained by us

**Comment by mrcdk** [comment]
Hi, for http://try-haxe.mrcdk.com I kinda hacked the docker support in (plus a few other things), it's not pretty but gets the job done :sweat_smile:

About improving try-haxe. In this thread https://groups.google.com/forum/#!topic/haxelang/z_o_iQkCoHs there are a few ideas but what most people wanted is basically a cloud IDE which clearly isn't the purpose of try-haxe so I'm not really sure in which direction improve it.

@andyli I haven't yet, I don't have much free time and there are a few problems. I'm still not sure if it has a free tier or it's only free the first 30 days. When you go to register it says 30 days free but then in the price tab the lower configuration says free without specifying any time limit. The other problem is that it seems to use its own docker image flavor https://console.ng.bluemix.net/docs/containers/container_index.html which looks like they are just renamed docker images but I haven't dive into the documentation yet. In any case, it seems that price won't be a problem.

**Comment by andyli** [comment]
> moved to our servers/clouds and maintained by us

To be clear, it can be hosted anywhere as long as the maintainer has access to it. There should be at least one more person in the HF who has access, but the service need not be hosted in an existing server. We don't want to "take over" the thing. Let's embrace the devops idea of "You Build It, You Run It".

I do want to have the source code repo moved to the HF account though, and we should give @mrcdk and @clemos admin access.

@mrcdk: I believe the free plan wouldn't expire after 30 days. But anyway, is your existing setup ready? If so, I can switch the DNS CNAME record to map try.haxe.org to try-haxe.mrcdk.com now. You may change the backend in whatever way as long as try-haxe.mrcdk.com point to the right location and we will be fine.

**Comment by markknol** [comment]
Doesnt such move mean that all current snippets are gone? I hope we can keep them all.

**Comment by clemos** [comment]
I can transfer the existing snippets of course ;)

**Comment by andyli** [comment]
@mrcdk Would you get in touch with @clemos to transfer the current snippets?

**Comment by mrcdk** [comment]
Right now is not the best moment for me, we're releasing a project in the next few weeks and it's kinda hectic around here. 

@andyli it's not ready yet, it needs a few small changes. I'll post here again once I manage to do them. Once the changes are done, I'll write down how to setup a server and submit a PR to clemos' repo (or wherever the official repo ends)
@clemos send me the snippets whenever you can, my e-mail is in my profile.

**Comment by andyli** [comment]
No problem. We will then keep using @clemos's server until you're free to handle the migration.

It would be better to sync the snippets when we actually migrate, because we don't want to lose the ones between now and the actually migration.

**Comment by clemos** [comment]
@andyli I'll see how we can do; I can probably setup an FTP access or something so one can `rsync` for a better migration.

**Comment by markknol** [comment]
@clemos @andyli what is the status of this?

**Comment by clemos** [comment]
I haven't taken the time to export the current try.haxe.org snippets yet, since as far as I understand, they're not critical in the transfer process right now.
It can be done very quickly when required, but I was kind of waiting for the target environment to be ready, so I don't work for nothing ;)

**Comment by andyli** [comment]
@mrcdk Do you feel comfortable to import the snippets now? ;)

**Comment by mrcdk** [comment]
Sorry guys, not the best moment right now. Maybe in a couple of weeks. I lost the computer and the VM I had the project set up in and I need to redo all the setup from scratch plus merge all the new updates to try.haxe.org which didn't look easy last time I checked.

**Comment by mrcdk** [comment]
Sorry for the huge delay I didn't have much time to work on this and it wasn't a easy job to get my fork up to date with the current try-haxe state. I had a couple of fail attempts before. My modifications are all over the place and I'm not sure how much longer I can keep them like this without fully rewriting big chunks of the project (if I ever find time)

Anyway, my fork is up to date now with a couple of issues related to linting.

I added a couple of things like more haxe versions and development versions too. New development versions are downloaded from here http://hxbuilds.s3-website-us-east-1.amazonaws.com/builds/haxe/linux64/ each 6 hours and will keep up to 10 versions removing the oldest one if there are more than 10. Scripts will default to the most recent stable version if the haxe version is not found in the server. Stable releases have to be added manually.

I had to remove the script mode because my fork uses multiple files for haxe code.

Now, about moving the snippets from try-haxe to my fork... I'm not a 100% sure how to do this without breaking one or the other. There will be id collisions for sure, I had them between my local server and my server.

@andyli @clemos

**Comment by gene-pavlovsky** [comment]
Was reading this with hope then noticed the date of the last comment. Is there any progress on this topic?
